velox/functions/prestosql/Zip.cpp

```diff
@@ -0,0 +1,221 @@
+
+    std::vector<BufferPtr> nestedResultIndices(numInputArrays);
+    std::vector<BufferPtr> nestedResultNulls(numInputArrays);
+    std::vector<vector_size_t*> rawNestedResultIndices(numInputArrays);
+    std::vector<uint64_t*> rawNestedResultNulls(numInputArrays);
+
+    for (int i = 0; i < numInputArrays; i++) {
+      nestedResultIndices[i] = allocateIndices(resultElementsSize, pool);
+      nestedResultNulls[i] = AlignedBuffer::allocate<bool>(
+          resultElementsSize, pool, bits::kNotNull);
+      rawNestedResultIndices[i] =
+          nestedResultIndices[i]->asMutable<vector_size_t>();
+      rawNestedResultNulls[i] = nestedResultNulls[i]->asMutable<uint64_t>();
+    }
+
+    // Final Row Vector
+    const auto finalArraySize = rows.end();
+    BufferPtr resultArrayIndices = allocateIndices(finalArraySize, pool);
+    BufferPtr resultArrayOffsets = allocateOffsets(finalArraySize, pool);
+    auto rawResultArrayOffsets = resultArrayOffsets->asMutable<vector_size_t>();
+
+    // Create right offsets/indexes for the individual and final result arrays.
+    int elementRow = 0;
+    rows.applyToSelected([&](auto row) {
+      // Get the max size for that row.
+      auto maxArraySize = maxArraySizes[row];
+      rawResultArrayOffsets[row] = elementRow;
+
+      for (int i = 0; i < numInputArrays; i++) {
+        auto offset = rawOffsets[i][indices[i][row]];
+        auto size = rawSizes[i][indices[i][row]];
+        std::iota(
+            rawNestedResultIndices[i] + elementRow,
+            rawNestedResultIndices[i] + elementRow + size,
+            offset);
+        bits::fillBits(
+            rawNestedResultNulls[i],
+            elementRow + size,
+            elementRow + maxArraySize,
+            false);
+      }
+      elementRow += maxArraySize;
+    });
+
+    // Now create result dictionary vectors.
+    std::vector<VectorPtr> resultDictionaryVectors(numInputArrays);
+
+    for (int i = 0; i < numInputArrays; i++) {
+      resultDictionaryVectors[i] = BaseVector::wrapInDictionary(
+          nestedResultNulls[i],
+          nestedResultIndices[i],
+          resultElementsSize,
+          baseVectors[i]->elements());
+    }
+
+    auto rowType = outputType->childAt(0);
+    auto rowVector = std::make_shared<RowVector>(
+        pool,
+        rowType,
+        BufferPtr(nullptr),
+        resultElementsSize,
+        resultDictionaryVectors);
+
+    // Now convert these to an Array
+    auto arrayVector = std::make_shared<ArrayVector>(
+        pool,
+        outputType,
+        BufferPtr(nullptr),
+        rows.end(),
+        resultArrayOffsets,
+        maxArraySizesBuffer,
+        std::move(rowVector));
+
+    context->moveOrCopyResult(arrayVector, rows, result);
+  }
+
+  static std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+    std::vector<std::shared_ptr<exec::FunctionSignature>> signatures;
+    static const auto kAritySize = (kMaxArity - kMinArity) + 1;
```

Comment:
This declaration should not interrupt the flow of the code. Move it to the very beginning of the method.